Use map of output streams, and return references to its elements

giacomofiorin opened the pull request:
As mentioned in the conversation for #481, update the registry of output files to use a `std::map` rather than two paired lists, which are no longer worth using any more.

giacomofiorin commented:
To summarize, I got confused a bit about which functions need to be overridden and which overloaded. The former gets a little tricky depending on the compiler. Eventually I fixed my mistakes and squashed them and the fixes back into one commit. 

Since we're dealing with a templated class (`colvar_grid<...>`) as a base, we're not really using base-class pointers and vtables, except in three cases (`value_input()`, `value_output()` and the destructor). But even those don't need to be `virtual`, either. I'd like to simplify this if possible, and instantiate them explicitly in the derived classes before moving on with this PR. Comments welcome, especially when concerning the ABF code.

jhenin commented on src/colvarproxy_io.cpp:
```diff
@@ -209,60 +213,43 @@ std::ostream * colvarproxy_io::output_stream(std::string const &output_name,
-
-  if (!(mode & (std::ios_base::app | std::ios_base::ate))) {
-    backup_file(output_name);
+  if (output_streams_.count(output_name) > 0) {
```
Could use output_stream_exists() ? There are a few instances of this.

giacomofiorin replied:
Done.

I also did the same for input streams, where the usage is very similar, even though it didn't come up in #481

jhenin commented on src/colvarbias_meta.cpp:
```diff
@@ -331,13 +330,12 @@ colvarbias_meta::~colvarbias_meta()
   colvarproxy *proxy = cvm::proxy;
 
-  if (proxy->get_output_stream(replica_hills_file)) {
+  if (proxy->output_stream_exists(replica_hills_file)) {
     proxy->close_output_stream(replica_hills_file);
   }
```
This could be done without testing for the existence first, since close_output_stream handles it gracefully.

giacomofiorin replied:
Done

jhenin commented on src/colvarmodule.cpp:
```diff
@@ -1092,17 +1084,18 @@ int colvarmodule::calc_scripted_forces()
-  if (cv_traj_os != NULL) {
+
+  if (proxy->output_stream_exists(cv_traj_name)) {
     // Take the opportunity to flush colvars.traj
-    proxy->flush_output_stream(cv_traj_os);
+    proxy->flush_output_stream(cv_traj_name);
   }
```
This could be done without testing if the flush_output_stream function was made more permissive, like close_output_stream, and ignored streams that aren't open. I'm not sure the test will be very effective at catching bugs, so I'd rather simplify the calling code.

giacomofiorin replied:
Done

jhenin commented on src/colvarproxy_io.cpp:
```diff
@@ -288,17 +277,29 @@ int colvarproxy_io::close_output_stream(std::string const &output_name)
+  for (std::map<std::string, std::ostream *>::iterator osi = output_streams_.begin();
+       osi != output_streams_.end();
+       osi++) {
+    (reinterpret_cast<std::ofstream *>(osi->second))->close();
```
How does this reinterpret_cast behave if the underlying object is not of type ofstream? Could that call an invalid function pointer and cause a crash?

giacomofiorin replied:
Since the map is private, the only chance that the wrong type of object can be stored there is if a derived proxy class overrides functions inconsistently. It's always possible, but it'd be a one-time error that would be caught when implementing that derived proxy class.

jhenin replied:
Wouldn't it be safer to use dynamic_cast for these?

giacomofiorin replied:
Ok, fair point. I was initially avoiding it because it is not as flexible with multiple inheritance, and we can't control what other classes will be used from here. But in the base one we do know that the inheritance is simple. Changing it.

giacomofiorin commented:
@jhenin I addressed your comment, feel free to look again but do keep in mind that this is still a draft.

Before finishing it, I would like to eliminate the last remaining virtual functions from the `colvar_grid` template class. Are you okay with that? Do the existing ABF tests cover a reasonable combination of features?

jhenin commented:
I think the tests should do the job, yes.

giacomofiorin commented:
@jhenin I have reconsidered switching from deriving the class template to instantiating an object of it. Replacing the many existing member functions with wrappers around the instantiated object would be tedious but conceptually simple.

But then there's the trickier issue that `read_raw` and `write_raw` need internally to call the _derived_ functions `value_input()` and `value_output()`. They need to do that because `colvar_grid_gradient` and `colvar_grid_count` are intertwined, and need to call each other at a relatively low level. Changing that design seems possible, but beyond scope for this PR.

Tests pass, I'll just update the Actions workflow because GitHub is nagging and wrap this up.
